### ISSUE TITLE
linux build fails because of template issue in third_party/webrtc/rtc_base/bitstream_reader.h

### ISSUE THREAD

**Issue by squid-f** [Eloston/ungoogled-chromium#1704]
**Describe the bug**
Linux build fails with the following errors:
```
In file included from ../../third_party/webrtc/rtc_base/bitstream_reader.cc:11:
../../third_party/webrtc/rtc_base/bitstream_reader.h:70:13: error: explicit specialization in non-namespace scope 'class webrtc::BitstreamReader'
   70 |   template <>
      |             ^
../../third_party/webrtc/rtc_base/bitstream_reader.h:71:29: error: template-id 'Read<bool>' in declaration of primary template
   71 |   ABSL_MUST_USE_RESULT bool Read<bool>() {
      |                             ^~~~~~~~~~
```

**To Reproduce**
1. Download source files
2. Run build using the RPM SPEC file successfully used to build Ungoogled-chromium 94.xxxx
3. Build fails at [10335/49258]

**Expected behavior**
Keep building

**Screenshots**
Full command line and related error:
```
[10335/49258] CXX obj/third_party/webrtc/rtc_base/bitstream_reader/bitstream_reader.o
FAILED: obj/third_party/webrtc/rtc_base/bitstream_reader/bitstream_reader.o 
x86_64-mageia-linux-gnu-g++ -MMD -MF obj/third_party/webrtc/rtc_base/bitstream_reader/bitstream_reader.o.d -DUSE_UDEV -DUSE_AURA=1 -DUSE_GLIB=1 -DUSE_NSS_CERTS=1 -DUSE_OZONE=1 -DUSE_X11=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -DNO_UNWIND_TABLES -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -DWEBRTC_ENABLE_PROTOBUF=1 -DRTC_ENABLE_VP9 -DWEBRTC_HAVE_SCTP -DENABLE_EXTERNAL_AUTH -DWEBRTC_USE_H264 -DHAVE_WEBRTC_VIDEO -DLOGGING_INSIDE_WEBRTC -DWEBRTC_LIBRARY_IMPL -DWEBRTC_NON_STATIC_TRACE_EVENT_HANDLERS=0 -DWEBRTC_CHROMIUM_BUILD -DWEBRTC_POSIX -DWEBRTC_LINUX -DABSL_ALLOCATOR_NOTHROW=1 -I../.. -Igen -I../../third_party/webrtc_overrides -I../../third_party/webrtc -Igen/third_party/webrtc -I../../third_party/abseil-cpp -fno-ident -fno-strict-aliasing --param=ssp-buffer-size=4 -fstack-protector -fno-unwind-tables -fno-asynchronous-unwind-tables -fPIC -pipe -pthread -Wno-builtin-macro-redefined -D__DATE__= -D__TIME__= -D__TIMESTAMP__= -Wall -Wno-unused-local-typedefs -Wno-deprecated-declarations -Wno-comments -Wno-packed-not-aligned -Wno-missing-field-initializers -Wno-unused-parameter -fno-omit-frame-pointer -fvisibility=hidden -std=gnu++14 -Wno-narrowing -Wno-class-memaccess -fno-exceptions -fno-rtti -fvisibility-inlines-hidden -O2 -pipe -Wformat -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector --param=ssp-buffer-size=4 -fasynchronous-unwind-tables -faligned-new -Wno-attributes -Wno-ignored-attributes -Wno-error=class-memaccess -Wno-error=unknown-pragmas -Wno-error=array-bounds -Wno-return-type -Wno-error=nonnull -Wno-error=free-nonheap-object -Wno-error=unused -Wno-error=unused-function -Wno-error=unused-variable -Wno-error=unused-but-set-variable -Wno-error=maybe-uninitialized -Wno-error=uninitialized -Wno-error=invalid-offsetof -Wno-error=stringop-overflow -std=gnu++17 -fpermissive -w -Wno-error=format-truncation -c ../../third_party/webrtc/rtc_base/bitstream_reader.cc -o obj/third_party/webrtc/rtc_base/bitstream_reader/bitstream_reader.o
In file included from ../../third_party/webrtc/rtc_base/bitstream_reader.cc:11:
../../third_party/webrtc/rtc_base/bitstream_reader.h:70:13: error: explicit specialization in non-namespace scope 'class webrtc::BitstreamReader'
   70 |   template <>
      |             ^
../../third_party/webrtc/rtc_base/bitstream_reader.h:71:29: error: template-id 'Read<bool>' in declaration of primary template
   71 |   ABSL_MUST_USE_RESULT bool Read<bool>() {
      |                             ^~~~~~~~~~
```

**Environment (please complete the following information):**
 - OS/Platform and version: Mageia 8 x64
 - ungoogled-chromium version: 95.0.4638.54
 - gcc version: output of gcc -v
 ```
 Utilisation des specs internes.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-mageia-linux-gnu/10/lto-wrapper
Cible : x86_64-mageia-linux-gnu
Configuré avec: ../configure --prefix=/usr --libexecdir=/usr/lib --with-slibdir=/lib64 --with-pkgversion='Mageia 10.3.0-2.mga8' --with-bugurl=https://bugs.mageia.org/ --mandir=/usr/share/man --infodir=/usr/share/info --enable-checking=release --enable-languages=c,c++,ada,fortran,objc,obj-c++,d --enable-linker-build-id --build=x86_64-mageia-linux-gnu --host=x86_64-mageia-linux-gnu --with-cpu=generic --with-system-zlib --enable-threads=posix --with-linker-hash-style=gnu --enable-shared --enable-long-long --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --enable-ssp --disable-libssp --disable-libffi --disable-werror --with-isl --with-python-dir=/lib/python3.8/site-packages --enable-lto --with-gcc-major-version-only
Modèle de thread: posix
Algorithmes de compression LTO supportés: zlib zstd
gcc version 10.3.0 (Mageia 10.3.0-2.mga8) 
```

**Additional context**
Hi
with the same spec file (beside patch adjustments to cope with code changes in version 95), I have successfully packaged 94.0.4606.81 and 94.0.4606.61
third_party/webrtc/rtc_base/bitstream_reader.cc seems a new addition with Chromium 95
My distribution has not packaged this 95 version yet. So, it might not be an ungoogled-chromium related issue.
However, before reporting a bug upstream to Chromium, I'd like to make sure it is unlikely related to ungoogled-chromium.
Plus, if you have some ideas how to solve this, it might save me some time ;)

Thanks!

**Comment by DAINRA** [comment]
Also had the problem with clang 13 when I was trying to get a build for v94 (-ffile-compilation-dir). Dunno if you are interested in building with clang, but Void fixed that by reversing this patch: [use-ffile-compilation-dir.patch](https://github.com/void-linux/void-packages/commit/7b272581e9ce52c164654df1b538c759b4092d10#diff-63c35dba045f439fc31a01b9f897653a95bc08bf486f0d7ab2e24f7a08012c72).
Void uses clang 12 but that should work for clang 11 also.
Sorry I can't be more helpful, but I don't follow chromium development that closely and I rely on Void's patches to build UC :/.

**Comment by squid-f** [comment]
Hi @DAINRA 
You helped a lot already as it is exactly the issue I get. I don’t know anything about Void. I’ll investigate it. 
@wchen342 sysroot cannot still be used as adding `build/linux/sysroot_scripts/install-sysroot.py --arch=amd64 ` leads to start downloading from a non existing web page. I will dig into your insights about `depot_tools`

**Comment by wchen342** [comment]
I will suggest you clone https://github.com/wchen342/ungoogled-chromium-building-scripts and run

```
./build.py init --shallow
./build.py sync --shallow --os linux
./build.py prepare --os linux
./build.py build -a x64 --os linux
```

It is designed exactly for this kind of scenario.

**Comment by squid-f** [comment]
Hi @wchen342 
In fact, I have been following the path you opened. I mean getting depo_tools and chromium source using hooks.
What I just did:
```
git clone https://github.com/Eloston/ungoogled-chromium.git
git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
export PATH="$PATH:`pwd`/depot_tools"
mkdir -p ./chromium && cd ./chromium
fetch --nohooks --no-history chromium
gclient runhooks
```
It looks like it ran well. 
Then, I did:
```
cd src
python3 -B path_to_unggoogled-chromium/utils/prune_binaries.py . \
  path_to_unggoogled-chromium/pruning.list
```
and it get this error message: `ERROR: 4631 files could not be pruned.`

I have been searching around for an explanation but without success. 
Might it be a version mismatch between Chromium and Ungoogled-chromium? 
I have been trying to force a version for Chromium but without success. I supposed by doing as above, I would get the latest and similar versions. But maybe not...

Any idea on what am I doing wrong??

**Comment by wchen342** [comment]
The prune error is normal. I have never seen this message gone, it has always been happening. I think it has something to do with different check-outs.

**Comment by squid-f** [comment]
Hum. I don’t get it when I do the pruning on the Chromium tarball source. 
I get this error when I get Chromium through depot_tools (fetch and gclient).

**Comment by wchen342** [comment]
I think the tarball has more extra files than the source tree. You can change L63 in `utils/_common.py` to `def get_logger(initial_level=logging.DEBUG):` to see which files are not pruned.

If those files doesn't exist then it shall be fine.

**Comment by wchen342** [comment]
@PF4Public @Ahrotahn Maybe we need two prune lists? I have noticed this for some time too since I almost always use source checkout instead of the google tarball and I always got this error.

**Comment by PF4Public** [comment]
> Maybe we need two prune lists? I have noticed this for some time too since I almost always use source checkout instead of the google tarball and I always got this error.

I suppose, the error comes from here:
https://github.com/Eloston/ungoogled-chromium/blob/c8011bab148961a16e7d89f68fb536cf57b4459b/utils/prune_binaries.py#L35-L37

Is it an error though? May we consider non-existing file as an _already pruned_ file and merely print a warning about that?

**Comment by Ahrotahn** [comment]
I think having two lists would require a little too much maintenance work.  One would need to download both sources every update and update four sets of lists along with that since the domain substitution list would need to be included as well.  Not to mention that the update scripts would need new options to signify which source set is being used since the checkout doesn't contain `*.hh` or `*.gyp*` files.  

The easiest solution might be to include a line in your script that runs the `update_lists.py` before doing the pruning and substitution.  

Is there a reason you prefer the source checkout over the tarball?  The vast majority of chromium builds use the tarball since it also contains release files like the pgo profiles and extra third-party code.  

@PF4Public It does seem a little silly to throw an error for files that don't exist, but I guess the intent is to alert the user that something has changed and requires a look into without letting them ignore the warnings and continue with a possibly compromised build.

**Comment by squid-f** [comment]
Hi. Great to have you on board as I have been pulling too many hairs compared to the one remaining :)

@Ahrotahn I am getting the source checkout because I need to get the clang package provided by sysroot. It looks like clang comes with the hooks. Is there a way to get it with the tarball instead ? 

However, for some reasons, the git checkout doesn't give me exactly the same files as in the tarball, for the same given Chromium version, of course (or I have made a mistake with my checkout... but gitk seems to tell me I have got the right version).

Now, the git checkout, still, miss an important folder compared to the tarball which is `tools/gn/build`
Hence, my build fails on `tools/gn/build/gen.py': [Errno 2] No such file or directory`
Reading the README.md in tools/gn, I run `python tools/gn/roll_gn.py`
But it seems deprecated as I get 
```
GN is autorolled by https://autoroll.skia.org/r/gn-chromium-autoroll
See also https://groups.google.com/a/chromium.org/g/build/c/2VH7DqyaVD0
```

Dead end again.

@wchen342 I cannot use your scripts as such, as I want to create a rpm package and not just build and run ungoogled chromium. I want also to apply the patches coming with my distro. 
I went through your patch and tried to get the source using, like you, git clone instead of fetch. But it doesn't help as I still don't get the gn/build folder... 
Actually, chromium doc alerts on the fact to use `fetch` and not `git clone`, but I have not seen any explanation why.

**Comment by PF4Public** [comment]
> I guess the intent is to alert the user that something has changed

I do agree, but this holds true _only_ if everyone uses tarballs, which turns to be a false assumption! That's why I suggested to make it non-fatal. Come to think of it, alternatively we may add a commandline key, say `--not-a-tarball`, which demotes this error into warning.

@squid-f I'm really sorry for us hijacking your issue.

> Actually, chromium doc alerts on the fact to use `fetch` and not `git clone`, but I have not seen any explanation why.

That is exactly your problem. Chromium comprises miriad of projects, every one of them needs to be checked out, that's wat fetch does!

**Comment by squid-f** [comment]
> 
> @squid-f I'm really sorry for us hijacking your issue.
> 
No worries @PF4Public    I am the one entertaining you with my lack of knowledge !

> > Actually, chromium doc alerts on the fact to use `fetch` and not `git clone`, but I have not seen any explanation why.
> 
> That is exactly your problem. Chromium comprises miriad of projects, every one of them needs to be checked out, that's what fetch does!

In fact, I do use `fetch` to check out the source. But I still don't get `tools/gn/build` folder and its configuration file. It is what does still block me after having passed all the other hurdles...

**Comment by wchen342** [comment]
There are two different ways of getting `gn`. If you are using tarball it is generated by running `bootstrap.py` but if you use source checkout then you need to run the one under `depot_tools`.

@Ahrotahn There are multiple reasons one needs to use source checkout. For example, for Android and Windows tarball doesn't work because they need files not existing in it. Also if you want to develop on top of chromium source tarball is not very convenient too.

**Comment by squid-f** [comment]
> There are two different ways of getting `gn`. If you are using tarball it is generated by running `bootstrap.py` but if you use source checkout then you need to run the one under `depot_tools`.

Sorry @wchen342 for my ignorance here. Could you be a bit more precise? There is no tools/gn folder coming with depot_tools.
Should I run `bootstrap_python3` from the depot_tools folder?

**Comment by wchen342** [comment]
No, you use the `gn` file under `depot_tools` you get in the very beginning where you also has `gclient`.

**Comment by squid-f** [comment]
> No, you use the `gn` file under `depot_tools` you get in the very beginning where you also has `gclient`.

So, I run the `gn` script included in depot_tools after fetching chromium. Is it what you mean?

**Comment by wchen342** [comment]
I think it will be easier if you can show the `spec` file. Basically in tarball, `gn` doesn't come with it so you need to run that `gen.py` script. But in the case of source checkout, `gn` comes bundled with `depot_tools` so where ever you call `gn` like `gn gen something`, you now use the `gn` in `depot_tools` directly.

**Comment by squid-f** [comment]
@wchen342 The spec is big. I put it in a termbin but I can paste it here if you prefer: https://termbin.com/gud8

The thing is that the command `python tools/gn/bootstrap/bootstrap.py --skip-generate-buildfiles '--gn-gen-args=use_sysroot=true ....` looks for `tools/gn/build/gen.py`
I don't get this gen.py by fetching chromium. I don't get it either by cloning depot_tools. Hence, why I am lost.

**Comment by squid-f** [comment]
@wchen342 complement: to get chromium, I run this script:
```
version=$(rpm -q --specfile --qf='%{version}\n' webylib.spec | head -n1)
echo "get depot_tools"
rm -rf depot_tools
git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
export PATH="$PATH:`pwd`/depot_tools"
export DEPOT_TOOLS_UPDATE=0
echo "fetch chromium"
mkdir -p ./chromium && cd ./chromium
fetch chromium
cd src
echo "Available tags"
git tag
echo "git checkout of .spec version"
git checkout $version
```

**Comment by wchen342** [comment]
Sorry some of the details slipped my mind since I am used to modify chromium sources. So what you need to do is:

1. Delete the line
```
python2 tools/gn/bootstrap/bootstrap.py  --skip-generate-buildfiles --gn-gen-args='%{gn_config}'
```

2. Delete
```
buildtools/linux64/gn
```
from `ungoogled-chromium/pruning.list`

3. Change
```
out/Release/gn gen --args='%{gn_config}' out/Release
```
to
```
third_party/depot_tools/gn gen --args='%{gn_config}' out/Release
```.

**Comment by squid-f** [comment]
Hi all. 
Here are some news on my endeavor :)

I'd like to thank you a lot, @wchen342 
You have helped me a lot and I have learned a lot as well.!

However, going through sysroot and getting the source from git, left me with a source folder of about 50Go. 
I might have not done it in the best way possible, though?
Anyway, this size of the source is not suitable to be used to build a package using my BS. I basically spend a lot of time to shuffle the source files around, before even starting to compile.

So, I spent some time to explore the option proposed by @DAINRA 
I have got also some inspiration from openSUSE community package to find the proper patches to be able to package Chromium using clang11 instead of clang13.
I don't think the spec of my distro was fully ready to build with clang and I had to tweak more things.

Long story short, I am now able to package Chromium 95 using the tarball source. I have then also packaged ungoogled-chromium on the top of it for Mageia. I have called this package WebyLib. Hope it is not an issue; I have contacted Eloston about this name, but I have never got any reply. Not sure my email has reached him.

Thanks all for your awesome support! :)
